From 11ae6f3d72f62b2767c83280c078f4eb74e18351 Mon Sep 17 00:00:00 2001
From: Derek Foreman <derek.foreman@collabora.com>
Date: Wed, 4 Sep 2024 06:49:56 -0500
Subject: [PATCH 208/213] [WIP-HDR]: drm/rockchip_vop2: add RK3588 plane color
 encoding and range properties

Add missing plane color properties. This is just a hack since the
selected values are not being programmed later inside the display
controller.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>
Signed-off-by: Nicolas Dufresne <nicolas.dufresne@collabora.com>
---
 drivers/gpu/drm/rockchip/rockchip_drm_vop2.c |  6 +++
 drivers/gpu/drm/rockchip/rockchip_drm_vop2.h |  4 ++
 drivers/gpu/drm/rockchip/rockchip_vop2_reg.c | 56 ++++++++++++++++++++
 3 files changed, 66 insertions(+)

diff --git a/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c b/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c
index 89fad5c35b18..9d34d6623cb6 100644
--- a/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c
+++ b/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c
@@ -2760,6 +2760,12 @@ static int vop2_plane_init(struct vop2 *vop2, struct vop2_win *win,
 		drm_plane_create_rotation_property(&win->base, DRM_MODE_ROTATE_0,
 						   DRM_MODE_ROTATE_0 |
 						   win->data->supported_rotations);
+	if (win->data->supported_color_encodings)
+		drm_plane_create_color_properties(&win->base,
+						  win->data->supported_color_encodings,
+						  win->data->supported_color_ranges,
+						  win->data->default_color_encoding,
+						  win->data->default_color_range);
 	drm_plane_create_alpha_property(&win->base);
 	drm_plane_create_blend_mode_property(&win->base, blend_caps);
 	drm_plane_create_zpos_property(&win->base, win->win_id, 0,
diff --git a/drivers/gpu/drm/rockchip/rockchip_drm_vop2.h b/drivers/gpu/drm/rockchip/rockchip_drm_vop2.h
index 910a4fdba5fd..feb8d3cc8fd9 100644
--- a/drivers/gpu/drm/rockchip/rockchip_drm_vop2.h
+++ b/drivers/gpu/drm/rockchip/rockchip_drm_vop2.h
@@ -189,6 +189,10 @@ struct vop2_win_data {
 	const u32 *formats;
 	const uint64_t *format_modifiers;
 	const unsigned int supported_rotations;
+	u32 supported_color_encodings;
+	enum drm_color_encoding default_color_encoding;
+	u32 supported_color_ranges;
+	enum drm_color_range default_color_range;
 
 	/**
 	 * @layer_sel_id: defined by register OVERLAY_LAYER_SEL or PORTn_LAYER_SEL
diff --git a/drivers/gpu/drm/rockchip/rockchip_vop2_reg.c b/drivers/gpu/drm/rockchip/rockchip_vop2_reg.c
index c44648e1bc0d..48c8ca000674 100644
--- a/drivers/gpu/drm/rockchip/rockchip_vop2_reg.c
+++ b/drivers/gpu/drm/rockchip/rockchip_vop2_reg.c
@@ -1129,6 +1129,13 @@ static const struct vop2_win_data rk3588_vop_win_data[] = {
 		.layer_sel_id = { 0, 0, 0, 0 },
 		.supported_rotations = DRM_MODE_ROTATE_90 | DRM_MODE_ROTATE_270 |
 				       DRM_MODE_REFLECT_X | DRM_MODE_REFLECT_Y,
+		.supported_color_encodings = BIT(DRM_COLOR_YCBCR_BT601) |
+					     BIT(DRM_COLOR_YCBCR_BT709) |
+					     BIT(DRM_COLOR_YCBCR_BT2020),
+		.default_color_encoding = DRM_COLOR_YCBCR_BT709,
+		.supported_color_ranges = BIT(DRM_COLOR_YCBCR_LIMITED_RANGE) |
+					  BIT(DRM_COLOR_YCBCR_FULL_RANGE),
+		.default_color_range = DRM_COLOR_YCBCR_LIMITED_RANGE,
 		.axi_bus_id = 0,
 		.axi_yrgb_r_id = 2,
 		.axi_uv_r_id = 3,
@@ -1148,6 +1155,13 @@ static const struct vop2_win_data rk3588_vop_win_data[] = {
 		.layer_sel_id = { 1, 1, 1, 1 },
 		.supported_rotations = DRM_MODE_ROTATE_90 | DRM_MODE_ROTATE_270 |
 				       DRM_MODE_REFLECT_X | DRM_MODE_REFLECT_Y,
+		.supported_color_encodings = BIT(DRM_COLOR_YCBCR_BT601) |
+					     BIT(DRM_COLOR_YCBCR_BT709) |
+					     BIT(DRM_COLOR_YCBCR_BT2020),
+		.default_color_encoding = DRM_COLOR_YCBCR_BT709,
+		.supported_color_ranges = BIT(DRM_COLOR_YCBCR_LIMITED_RANGE) |
+					  BIT(DRM_COLOR_YCBCR_FULL_RANGE),
+		.default_color_range = DRM_COLOR_YCBCR_LIMITED_RANGE,
 		.type = DRM_PLANE_TYPE_PRIMARY,
 		.axi_bus_id = 0,
 		.axi_yrgb_r_id = 6,
@@ -1167,6 +1181,13 @@ static const struct vop2_win_data rk3588_vop_win_data[] = {
 		.layer_sel_id = { 4, 4, 4, 4 },
 		.supported_rotations = DRM_MODE_ROTATE_90 | DRM_MODE_ROTATE_270 |
 				       DRM_MODE_REFLECT_X | DRM_MODE_REFLECT_Y,
+		.supported_color_encodings = BIT(DRM_COLOR_YCBCR_BT601) |
+					     BIT(DRM_COLOR_YCBCR_BT709) |
+					     BIT(DRM_COLOR_YCBCR_BT2020),
+		.default_color_encoding = DRM_COLOR_YCBCR_BT709,
+		.supported_color_ranges = BIT(DRM_COLOR_YCBCR_LIMITED_RANGE) |
+					  BIT(DRM_COLOR_YCBCR_FULL_RANGE),
+		.default_color_range = DRM_COLOR_YCBCR_LIMITED_RANGE,
 		.type = DRM_PLANE_TYPE_PRIMARY,
 		.axi_bus_id = 1,
 		.axi_yrgb_r_id = 2,
@@ -1186,6 +1207,13 @@ static const struct vop2_win_data rk3588_vop_win_data[] = {
 		.layer_sel_id =  { 5, 5, 5, 5 },
 		.supported_rotations = DRM_MODE_ROTATE_90 | DRM_MODE_ROTATE_270 |
 				       DRM_MODE_REFLECT_X | DRM_MODE_REFLECT_Y,
+		.supported_color_encodings = BIT(DRM_COLOR_YCBCR_BT601) |
+					     BIT(DRM_COLOR_YCBCR_BT709) |
+					     BIT(DRM_COLOR_YCBCR_BT2020),
+		.default_color_encoding = DRM_COLOR_YCBCR_BT709,
+		.supported_color_ranges = BIT(DRM_COLOR_YCBCR_LIMITED_RANGE) |
+					  BIT(DRM_COLOR_YCBCR_FULL_RANGE),
+		.default_color_range = DRM_COLOR_YCBCR_LIMITED_RANGE,
 		.type = DRM_PLANE_TYPE_PRIMARY,
 		.axi_bus_id = 1,
 		.axi_yrgb_r_id = 6,
@@ -1204,6 +1232,13 @@ static const struct vop2_win_data rk3588_vop_win_data[] = {
 		.base = 0x1800,
 		.layer_sel_id = { 2, 2, 2, 2 },
 		.supported_rotations = DRM_MODE_REFLECT_Y,
+		.supported_color_encodings = BIT(DRM_COLOR_YCBCR_BT601) |
+					     BIT(DRM_COLOR_YCBCR_BT709) |
+					     BIT(DRM_COLOR_YCBCR_BT2020),
+		.default_color_encoding = DRM_COLOR_YCBCR_BT709,
+		.supported_color_ranges = BIT(DRM_COLOR_YCBCR_LIMITED_RANGE) |
+					  BIT(DRM_COLOR_YCBCR_FULL_RANGE),
+		.default_color_range = DRM_COLOR_YCBCR_LIMITED_RANGE,
 		.type = DRM_PLANE_TYPE_OVERLAY,
 		.axi_bus_id = 0,
 		.axi_yrgb_r_id = 0x0a,
@@ -1221,6 +1256,13 @@ static const struct vop2_win_data rk3588_vop_win_data[] = {
 		.base = 0x1a00,
 		.layer_sel_id = { 3, 3, 3, 3 },
 		.supported_rotations = DRM_MODE_REFLECT_Y,
+		.supported_color_encodings = BIT(DRM_COLOR_YCBCR_BT601) |
+					     BIT(DRM_COLOR_YCBCR_BT709) |
+					     BIT(DRM_COLOR_YCBCR_BT2020),
+		.default_color_encoding = DRM_COLOR_YCBCR_BT709,
+		.supported_color_ranges = BIT(DRM_COLOR_YCBCR_LIMITED_RANGE) |
+					  BIT(DRM_COLOR_YCBCR_FULL_RANGE),
+		.default_color_range = DRM_COLOR_YCBCR_LIMITED_RANGE,
 		.type = DRM_PLANE_TYPE_OVERLAY,
 		.axi_bus_id = 0,
 		.axi_yrgb_r_id = 0x0c,
@@ -1238,6 +1280,13 @@ static const struct vop2_win_data rk3588_vop_win_data[] = {
 		.format_modifiers = format_modifiers,
 		.layer_sel_id =  { 6, 6, 6, 6 },
 		.supported_rotations = DRM_MODE_REFLECT_Y,
+		.supported_color_encodings = BIT(DRM_COLOR_YCBCR_BT601) |
+					     BIT(DRM_COLOR_YCBCR_BT709) |
+					     BIT(DRM_COLOR_YCBCR_BT2020),
+		.default_color_encoding = DRM_COLOR_YCBCR_BT709,
+		.supported_color_ranges = BIT(DRM_COLOR_YCBCR_LIMITED_RANGE) |
+					  BIT(DRM_COLOR_YCBCR_FULL_RANGE),
+		.default_color_range = DRM_COLOR_YCBCR_LIMITED_RANGE,
 		.type = DRM_PLANE_TYPE_OVERLAY,
 		.axi_bus_id = 1,
 		.axi_yrgb_r_id = 0x0a,
@@ -1255,6 +1304,13 @@ static const struct vop2_win_data rk3588_vop_win_data[] = {
 		.base = 0x1e00,
 		.layer_sel_id =  { 7, 7, 7, 7 },
 		.supported_rotations = DRM_MODE_REFLECT_Y,
+		.supported_color_encodings = BIT(DRM_COLOR_YCBCR_BT601) |
+					     BIT(DRM_COLOR_YCBCR_BT709) |
+					     BIT(DRM_COLOR_YCBCR_BT2020),
+		.default_color_encoding = DRM_COLOR_YCBCR_BT709,
+		.supported_color_ranges = BIT(DRM_COLOR_YCBCR_LIMITED_RANGE) |
+					  BIT(DRM_COLOR_YCBCR_FULL_RANGE),
+		.default_color_range = DRM_COLOR_YCBCR_LIMITED_RANGE,
 		.type = DRM_PLANE_TYPE_OVERLAY,
 		.axi_bus_id = 1,
 		.axi_yrgb_r_id = 0x0c,
-- 
2.43.0

